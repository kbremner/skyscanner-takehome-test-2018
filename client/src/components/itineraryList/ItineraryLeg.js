import React from 'react';
import BpkText from 'bpk-component-text';
import BpkSmallLongArrowRight from 'bpk-component-icon/sm/long-arrow-right';
import PropTypes from '../../types';
import './ItineraryLeg.scss';

const formatDuration = duration => {
  const numHours = Math.floor(duration / 60);
  const numMinutes = duration % 60;
  if (numMinutes === 0) {
    return `${numHours}h`;
  }
  if (numHours === 0) {
    return `${numMinutes}m`;
  }
  return `${numHours}h ${numMinutes}`;
}

const formatNumStops = numStops => numStops === 0
  ? 'Direct'
  : `${numStops} Stop${numStops === 1 ? '' : 's'}`;

const formatTime = dateTimeStr => {
  const dateTime = new Date(dateTimeStr);
  return [dateTime.getHours(), dateTime.getMinutes()]
    .map(timeComponent => `${timeComponent < 10 ? '0' : ''}${timeComponent}`)
    .join(':');
}

const Segment = ({ dateTime, code }) => (
  <div className="itinerary-leg__segment">
    <BpkText>{formatTime(dateTime)}</BpkText>
    <BpkText className="itinerary-leg__location">{code}</BpkText>
  </div>
);

const ItineraryLeg = ({ className, carrier, segments, duration, numStops }) => (
  <div className={[className, 'itinerary-leg'].join(' ')}>
    <div>
      <img
        className="itinerary-leg__airline-logo"
        alt={`${carrier.name} logo`}
        src={carrier.faviconUrl}
      />
    </div>
    
    <div className="itinerary-leg__segments">
      {segments.map(({ origin, destination }, i, allSegments) => {
        const elems = [
          <Segment dateTime={origin.dateTime} code={origin.code} key={origin.code} />,
          <BpkSmallLongArrowRight className="itinerary-leg__arrow" key={`arrow-${i}`} />
        ];
        if (i === (allSegments.length - 1)) {
          // last segment, include destination
          elems.push(<Segment dateTime={destination.dateTime} code={destination.code} key={destination.code} />);
        }
        return elems;
      })}
    </div>

    <div className="itinerary-leg__summary">
      <BpkText className="itinerary-leg__duration">{formatDuration(duration)}</BpkText>
      <BpkText className="itinerary-leg__num-stops">{formatNumStops(numStops)}</BpkText>
    </div>
  </div>
);

ItineraryLeg.propTypes = {
  carrier: PropTypes.carrier.isRequired,
  segments: PropTypes.arrayOf(PropTypes.segment).isRequired,
  duration: PropTypes.number.isRequired,
  numStops: PropTypes.number.isRequired,
  className: PropTypes.string,
};

ItineraryLeg.defaultProps = {
  className: null
};

export default ItineraryLeg;