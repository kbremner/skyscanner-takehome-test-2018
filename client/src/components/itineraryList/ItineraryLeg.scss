@import "~bpk-mixins";

.itinerary-leg {
  padding-bottom: $bpk-spacing-md;

  .itinerary-leg__airline-logo {
    border-radius: $bpk-border-radius-sm;
    height: 30px;
  }

  &, .itinerary-leg__segments {
    display: flex;
    align-items: center;
  }

  .itinerary-leg__segments {
    flex-grow: 1;
    flex-wrap: wrap;
  }

  .itinerary-leg__segment, .itinerary-leg__summary {
    display: flex;
    flex-direction: column;
  }

  .itinerary-leg__summary {
    align-items: flex-end;
    flex-shrink: 0;
  }

  .itinerary-leg__segment {
    align-items: flex-start;
    padding: 0 $bpk-spacing-sm;
  }

  .itinerary-leg__location, .itinerary-leg__duration {
    color: $bpk-color-gray-300;
  }
  
  .itinerary-leg__arrow {
    fill: $bpk-color-gray-300;
  }

  .itinerary-leg__num-stops {
    color: $bpk-color-green-500;
  }
}