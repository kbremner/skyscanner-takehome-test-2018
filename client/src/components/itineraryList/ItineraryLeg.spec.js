import React from 'react';
import { shallow } from 'enzyme';
import ItineraryLeg from './ItineraryLeg';

const defaultProps = () => ({
  id: '1',
  carrier: {
    name: 'easyjet',
    code: 'EZ'
  },
  duration: 210,
  numStops: 0,
  segments: [11, 12, 13, 14].map(val => ({
    origin: {
      dateTime: `${val}:00`, code: `CODE-${val}`
    },
    destination: {
      dateTime: `${val+1}:00`, code: `CODE-${val+1}`
    }
  }))
});

const setup = (overrides = {}) => {
  const props = {
    ...defaultProps(),
    ...overrides
  };

  const wrapper = shallow(<ItineraryLeg {...props} />);

  return { wrapper, props };
};

describe('ItineraryLeg', () => {
  it('renders', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('adds class name to card provided', () => {
    const { wrapper, props } = setup({ className: 'test-class-name' });
    expect(wrapper.hasClass(props.className)).toBe(true);
  });

  it('renders just minutes when duration is less than 1h', () => {
    const { wrapper } = setup({ duration: 59 });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders just hours when duration has no minutes component', () => {
    const { wrapper } = setup({ duration: 180 });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders stops correctly when just 1', () => {
    const { wrapper } = setup({ numStops: 1 });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders stops correctly when more than 1', () => {
    const { wrapper } = setup({ numStops: 2 });
    expect(wrapper).toMatchSnapshot();
  });
});
