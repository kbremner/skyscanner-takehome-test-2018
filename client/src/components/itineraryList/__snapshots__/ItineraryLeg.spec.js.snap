// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItineraryLeg renders 1`] = `
<div
  className=" itinerary-leg"
>
  <div>
    <img
      alt="easyjet logo"
      className="itinerary-leg__airline-logo"
    />
  </div>
  <div
    className="itinerary-leg__segments"
  >
    <Segment
      code="CODE-11"
      dateTime="11:00"
      key="CODE-11"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-0"
    />
    <Segment
      code="CODE-12"
      dateTime="12:00"
      key="CODE-12"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-1"
    />
    <Segment
      code="CODE-13"
      dateTime="13:00"
      key="CODE-13"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-2"
    />
    <Segment
      code="CODE-14"
      dateTime="14:00"
      key="CODE-14"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-3"
    />
    <Segment
      code="CODE-15"
      dateTime="15:00"
      key="CODE-15"
    />
  </div>
  <div
    className="itinerary-leg__summary"
  >
    <BpkText
      bold={false}
      className="itinerary-leg__duration"
      tagName="span"
      textStyle="base"
    >
      3h 30
    </BpkText>
    <BpkText
      bold={false}
      className="itinerary-leg__num-stops"
      tagName="span"
      textStyle="base"
    >
      Direct
    </BpkText>
  </div>
</div>
`;

exports[`ItineraryLeg renders just hours when duration has no minutes component 1`] = `
<div
  className=" itinerary-leg"
>
  <div>
    <img
      alt="easyjet logo"
      className="itinerary-leg__airline-logo"
    />
  </div>
  <div
    className="itinerary-leg__segments"
  >
    <Segment
      code="CODE-11"
      dateTime="11:00"
      key="CODE-11"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-0"
    />
    <Segment
      code="CODE-12"
      dateTime="12:00"
      key="CODE-12"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-1"
    />
    <Segment
      code="CODE-13"
      dateTime="13:00"
      key="CODE-13"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-2"
    />
    <Segment
      code="CODE-14"
      dateTime="14:00"
      key="CODE-14"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-3"
    />
    <Segment
      code="CODE-15"
      dateTime="15:00"
      key="CODE-15"
    />
  </div>
  <div
    className="itinerary-leg__summary"
  >
    <BpkText
      bold={false}
      className="itinerary-leg__duration"
      tagName="span"
      textStyle="base"
    >
      3h
    </BpkText>
    <BpkText
      bold={false}
      className="itinerary-leg__num-stops"
      tagName="span"
      textStyle="base"
    >
      Direct
    </BpkText>
  </div>
</div>
`;

exports[`ItineraryLeg renders just minutes when duration is less than 1h 1`] = `
<div
  className=" itinerary-leg"
>
  <div>
    <img
      alt="easyjet logo"
      className="itinerary-leg__airline-logo"
    />
  </div>
  <div
    className="itinerary-leg__segments"
  >
    <Segment
      code="CODE-11"
      dateTime="11:00"
      key="CODE-11"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-0"
    />
    <Segment
      code="CODE-12"
      dateTime="12:00"
      key="CODE-12"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-1"
    />
    <Segment
      code="CODE-13"
      dateTime="13:00"
      key="CODE-13"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-2"
    />
    <Segment
      code="CODE-14"
      dateTime="14:00"
      key="CODE-14"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-3"
    />
    <Segment
      code="CODE-15"
      dateTime="15:00"
      key="CODE-15"
    />
  </div>
  <div
    className="itinerary-leg__summary"
  >
    <BpkText
      bold={false}
      className="itinerary-leg__duration"
      tagName="span"
      textStyle="base"
    >
      59m
    </BpkText>
    <BpkText
      bold={false}
      className="itinerary-leg__num-stops"
      tagName="span"
      textStyle="base"
    >
      Direct
    </BpkText>
  </div>
</div>
`;

exports[`ItineraryLeg renders stops correctly when just 1 1`] = `
<div
  className=" itinerary-leg"
>
  <div>
    <img
      alt="easyjet logo"
      className="itinerary-leg__airline-logo"
    />
  </div>
  <div
    className="itinerary-leg__segments"
  >
    <Segment
      code="CODE-11"
      dateTime="11:00"
      key="CODE-11"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-0"
    />
    <Segment
      code="CODE-12"
      dateTime="12:00"
      key="CODE-12"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-1"
    />
    <Segment
      code="CODE-13"
      dateTime="13:00"
      key="CODE-13"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-2"
    />
    <Segment
      code="CODE-14"
      dateTime="14:00"
      key="CODE-14"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-3"
    />
    <Segment
      code="CODE-15"
      dateTime="15:00"
      key="CODE-15"
    />
  </div>
  <div
    className="itinerary-leg__summary"
  >
    <BpkText
      bold={false}
      className="itinerary-leg__duration"
      tagName="span"
      textStyle="base"
    >
      3h 30
    </BpkText>
    <BpkText
      bold={false}
      className="itinerary-leg__num-stops"
      tagName="span"
      textStyle="base"
    >
      1 Stop
    </BpkText>
  </div>
</div>
`;

exports[`ItineraryLeg renders stops correctly when more than 1 1`] = `
<div
  className=" itinerary-leg"
>
  <div>
    <img
      alt="easyjet logo"
      className="itinerary-leg__airline-logo"
    />
  </div>
  <div
    className="itinerary-leg__segments"
  >
    <Segment
      code="CODE-11"
      dateTime="11:00"
      key="CODE-11"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-0"
    />
    <Segment
      code="CODE-12"
      dateTime="12:00"
      key="CODE-12"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-1"
    />
    <Segment
      code="CODE-13"
      dateTime="13:00"
      key="CODE-13"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-2"
    />
    <Segment
      code="CODE-14"
      dateTime="14:00"
      key="CODE-14"
    />
    <Component
      className="itinerary-leg__arrow"
      key="arrow-3"
    />
    <Segment
      code="CODE-15"
      dateTime="15:00"
      key="CODE-15"
    />
  </div>
  <div
    className="itinerary-leg__summary"
  >
    <BpkText
      bold={false}
      className="itinerary-leg__duration"
      tagName="span"
      textStyle="base"
    >
      3h 30
    </BpkText>
    <BpkText
      bold={false}
      className="itinerary-leg__num-stops"
      tagName="span"
      textStyle="base"
    >
      2 Stops
    </BpkText>
  </div>
</div>
`;
